sky/serve/autoscalers.py

```diff
@@ -192,5 +199,200 @@ def evaluate_scaling(
+        return {'use_spot': False, 'spot_recovery': None}
+
+    def _get_desired_num_replicas(self, num_ready_replicas: int) -> int:
+        # Convert to requests per second.
+        num_requests_per_second = len(
+            self.request_timestamps) / self.rps_window_size
+        # Edge case: num_replicas is zero.
+        requests_per_replica = (num_requests_per_second / num_ready_replicas if
+                                num_ready_replicas else num_requests_per_second)
+        target_num_replicas = math.ceil(requests_per_replica /
+                                        self.target_qps_per_replica)
+        target_num_replicas = max(self.min_replicas,
+                                  min(self.max_replicas, target_num_replicas))
+        logger.info(f'Requests per replica: {requests_per_replica}, '
+                    f'Current target number of replicas: {target_num_replicas}')
+
+        if target_num_replicas > self.target_num_replicas:
+            self.upscale_counter += 1
+            self.downscale_counter = 0
+            if self.upscale_counter >= self.scale_up_consecutive_periods:
+                self.upscale_counter = 0
+                return target_num_replicas
+        elif target_num_replicas < self.target_num_replicas:
+            self.downscale_counter += 1
+            self.upscale_counter = 0
+            if self.downscale_counter >= self.scale_down_consecutive_periods:
+                self.downscale_counter = 0
+                return target_num_replicas
+        else:
+            self.upscale_counter = self.downscale_counter = 0
+        return self.target_num_replicas
+
+    def handle_preemption_history(self, history: List[str]) -> None:
+        for zone in history:
+            self.spot_placer.handle_preemption(zone)
+
+    def evaluate_scaling(
+        self,
+        replica_infos: List['replica_managers.ReplicaInfo'],
+    ) -> List[AutoscalerDecision]:
+        current_time = time.time()
+        # TODO(tian): Consider non-alive replicas.
+        alive_replica_infos = [info for info in replica_infos if info.is_alive]
+        num_replicas = len(alive_replica_infos)
+        num_ready_replicas = len([
+            info for info in replica_infos
+            if info.status == serve_state.ReplicaStatus.READY
+        ])
+
+        # Check if cooldown period has passed since the last scaling operation.
+        # Only cooldown if bootstrapping is done.
+        if num_replicas >= self.min_replicas:
+            if current_time - self.last_scale_operation < self.cooldown:
+                logger.info(
+                    f'Current time: {current_time}, '
+                    f'last scale operation: {self.last_scale_operation}, '
+                    f'cooldown: {self.cooldown}')
+                logger.info('Cooldown period has not passed since last scaling '
+                            'operation. Skipping scaling.')
+                return []
+        else:
+            # Bootstrap.
```

Comment:
`num_replicas < self.min_replicas` might happen when many spot instances are preempted and not during bootstrap. 
E.g. min_replicas = 2, num_replicas = 1, target_num_replicas = 3.

Reply:
Good point! Now use a flag to indicate initialized or not.